### PR TITLE
Volumes grpc interface

### PR DESCRIPTION
**Pull request by gtardif**
**What I did**
Replace #640 (including commit from @ulyssessouza) to follow up with things we did pair programming and fixes in metrics)

**Related issue**
https://github.com/docker/compose-cli/issues/604

**(not mandatory) A picture of a cute animal, if possible in relation with what you did**

### CODE REVIEW

**Comment by gtardif** [comment]
CI workflows need to be fixed to include build happening on fork ([for this PR](https://github.com/gtardif/compose-cli/tree/volumes-grpc))

**Comment by gtardif** [comment]
This will require a follow up PR in node-sdk to add volumes API, and add a volume sample (already tested, working on ACI context)

File: cli/cmd/serve.go [hunk 1]
```diff
@@ -64,6 +65,7 @@ func runServe(ctx context.Context, opts serveOpts) error {
 	p := proxy.New(ctx)
 
 	containersv1.RegisterContainersServer(s, p)
+	volumesv1.RegisterVolumesServer(s, p)
```
Comment (chris-crone) [inline cli/cmd/serve.go:68]: nit: Any reason not to sort these?

File: protos/volumes/v1/volumes.pb.go [hunk 2]
```diff
@@ -0,0 +1,855 @@
+//
+//  Copyright 2020 Docker, Inc.
```
Comment (chris-crone) [inline protos/volumes/v1/volumes.pb.go:2]: This can't be Docker. Inc. with just a DCO:

```suggestion
//  Copyright 2020 Docker Compose CLI authors
```

File: protos/volumes/v1/volumes.proto [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	rpc VolumesDelete(VolumesDeleteRequest) returns (VolumesDeleteResponse);
+}
+
+message Volume {
+	string id = 1;
+	string description = 2;
+}
+
+message AciVolumeCreateOptions {
+	string storage_account= 1;
```
Comment (chris-crone) [inline protos/volumes/v1/volumes.proto:36]:
```suggestion
	string storage_account = 1;
```

File: protos/volumes/v1/volumes.proto [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+message VolumesListRequest {
+}
+
+message VolumesListResponse {
+	repeated Volume volumes = 1;
+}
+
+message VolumesDeleteRequest {
+	string id = 1;
+	repeated google.protobuf.Any options = 2;
```
Comment (chris-crone) [inline protos/volumes/v1/volumes.proto:59]: Curious what options we can pass here?

File: server/proxy/volumes.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// VolumesCreate creates a volume.
+func (p *proxy) VolumesCreate(ctx context.Context, req *volumesv1.VolumesCreateRequest) (*volumesv1.VolumesCreateResponse, error) {
+	backendClient := Client(ctx)
+	contextType := ""
+	if backendClient != nil {
+		contextType = backendClient.ContextType()
+	}
+
+	if contextType != "aci" {
```
Comment (chris-crone) [inline server/proxy/volumes.go:36]: Do we have consts for the types that we can use?

File: server/proxy/volumes.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		return nil, fmt.Errorf("volume API only available in ACI context")
```
Comment (chris-crone) [inline server/proxy/volumes.go:37]: You can use not implemented from our errdefs:
```suggestion
		return nil, errors.Wrapf(errdefs.ErrNotImplemented, "volume API only available in ACI context")
```

**Review by chris-crone — COMMENTED** [review]
Generally looks good, only minor issues

File: cli/cmd/serve.go (hunk 1 above)
Comment (gtardif) [reply cli/cmd/serve.go]: no, fixed

File: protos/volumes/v1/volumes.proto (hunk 4 above)
Comment (gtardif) [reply protos/volumes/v1/volumes.proto]: This was to potentially pass more info for other backends. but we can add this later as an optional field, this wil not break the API. removing this field

File: server/proxy/volumes.go (hunk 5 above)
Comment (gtardif) [reply server/proxy/volumes.go]: Yes, Thx !

File: server/proxy/volumes.go (hunk 6 above)
Comment (gtardif) [reply server/proxy/volumes.go]: fixed

File: server/proxy/volumes.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+// VolumesCreate creates a volume.
+func (p *proxy) VolumesCreate(ctx context.Context, req *volumesv1.VolumesCreateRequest) (*volumesv1.VolumesCreateResponse, error) {
+	backendClient := Client(ctx)
+	contextType := ""
+	if backendClient != nil {
+		contextType = backendClient.ContextType()
+	}
+
+	if contextType != store.AciContextType {
```
Comment (rumpl) [inline server/proxy/volumes.go:39]: Not sure we really want this here, the backend should return a `not implemented` error instead and then the cli can say that the volumes is not implemented. no ?

Comment (gtardif) [reply server/proxy/volumes.go:39]: indeed this is not reached if not ACI

File: protos/streams/v1/streams.pb.go [hunk 8]
```diff
@@ -23,15 +23,16 @@ package v1
 
 import (
 	context "context"
+	reflect "reflect"
```
Comment (rumpl) [inline protos/streams/v1/streams.pb.go:26]: Why did these change? How did you generate the protos?

File: server/proxy/volumes.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		return nil, errors.Wrapf(errdefs.ErrNotImplemented, "volume API only available in ACI context")
+	}
+	aciReq := req.GetAciOption()
+	if aciReq == nil {
```
Comment (rumpl) [inline server/proxy/volumes.go:43]: This check should be in the backend too no?

Comment (gtardif) [reply server/proxy/volumes.go:32]: Actually we need to read the gRPC param and convert it into params for the backend. We can just pass an empty storage account by default and let the backend return the error, that would avoid different errors for different cases, yes.

File: protos/streams/v1/streams.pb.go (hunk 8 above)
Comment (gtardif) [reply protos/streams/v1/streams.pb.go:26]: This is me running `goimports -local github.com/docker/compose-cli -w .` after `make protos`. I'll remove this to not change all generated files in this PR, only volume changes required

File: server/metrics.go [hunk 10]
```diff
@@ -36,9 +36,12 @@ var (
 		"/com.docker.api.protos.containers.v1.Containers/Kill":    "kill",
 		"/com.docker.api.protos.containers.v1.Containers/Inspect": "inspect",
 		"/com.docker.api.protos.containers.v1.Containers/Logs":    "logs",
-		"/com.docker.api.protos.streams.v1.Streaming/NewStream":   "",
+		"/com.docker.api.protos.streams.v1.Streaming/NewStream":   "streaming",
```
Comment (rumpl) [inline server/metrics.go:39]: Why this change?

Comment (gtardif) [reply server/metrics.go:39]: I just didn't want to let this empty, if anyone calls this, we'll see something we can follow in the metrics

**Review by rumpl — APPROVED** [review]
LGTM
